keps/sig-etcd/4647-cluster-feature-gate/README.md

```diff
@@ -0,0 +1,468 @@
+```go
+if s.FeatureEnabled(FeatureA) && s.ClusterVersion() <= "3.6" {implementation 1}
+if s.FeatureEnabled(FeatureA) && s.ClusterVersion() >= "3.7" {implementation 2}
+```
+This way the cluster would work consistently because there is a single ClusterVersion across the whole cluster.
+
+## Design Details
+
+On high level, a cluster feature gate would need:
+1. a way to [register new feature gates](#register-new-feature-gates) when a developer adds new features.
+1. a way to [set the feature gates](#set-the-feature-gates) when each member starts up.
+1. a [consensus algorithm](#consensus-algorithm) to determine if a feature is enabled for the whole cluster, even in mixed version scenarios.
+1. an [interface](#interface-change-for-raftcluster) to query if a feature is enabled for the whole cluster to be used in the code.
+1. [client APIs](#client-apis-changes) to query if a feature is enabled for the whole cluster.
+
+### Register New Feature Gates
+
+A feature can be registered as server level feature or cluster level feature, but not both.
+For data compatibility, we do not support changing a feature gate from one type to another. 
+If a developer finds out that their feature needs to be a cluster level feature instead of server level after a release, they should add a new feature and use a new feature name.
+
+```go
+DefaultEtcdServerFeatureGates := map[Feature]VersionedSpecs {
+		featureA: VersionedSpecs{
+			{Version: mustParseVersion("3.6"), Default: false, PreRelease: Beta},
+			{Version: mustParseVersion("3.7"), Default: true, PreRelease: GA},
+		},
+		featureB: VersionedSpecs{
+			{Version: mustParseVersion("3.7"), Default: false, PreRelease: Alpha},
+		},
+}
+DefaultEtcdClusterFeatureGates := map[Feature]VersionedSpecs {
+		featureC: VersionedSpecs{
+			{Version: mustParseVersion("3.7"), Default: false, PreRelease: Beta},
+		},
+		featureD: VersionedSpecs{
+			{Version: mustParseVersion("3.6"), Default: false, PreRelease: Alpha},
+			{Version: mustParseVersion("3.7"), Default: true, PreRelease: Deprecated},
+		}
+}
+```
+
+### Set the Feature Gates
+Conceptually, a feature is a new functionality etcd provides, and it has to work on individual servers first before working on the cluster. The main difference between server level and cluster level features is whether their usage require consistent value across all members.
+
+When a user start up a new etcd cluster, it is unnecessary and confusing to require them to know the difference between server level and cluster level features. It is really up to the developers to register them to the proper type in the code.
+
+So the cluster feature gates for a server can be set in the same way as the server level feature gates: with the `--feature-gates`flag or in the `config-file` during startup. We do not support dynamically changing the feature gates when the server is running.
```

Comment:
I would prefer to separate configuration of server and cluster features in different command line flag. The reason is user expectation. Users expect consistent behavior of flags, are they enabled immediately or they require to reconfigure the whole cluster. They shouldn't need to check the type of flag. 

What do you think?

Reply:
makes sense. Modified the section.